django/template/loader_tags.py

```diff
@@ -178,12 +178,22 @@ def __init__(self, template, *args, **kwargs):
         super(IncludeNode, self).__init__(*args, **kwargs)
 
     def render(self, context):
+        """
+        Render the specified template and context. Cache the template object
+        in render_context to avoid reparsing and loading when used in a for
+        loop.
+        """
         try:
             template = self.template.resolve(context)
             # Does this quack like a Template?
             if not callable(getattr(template, 'render', None)):
-                # If not, we'll try get_template
-                template = context.template.engine.get_template(template)
+                # If not, we'll try our cache, and get_template
+                template_name = template
+                template = context.render_context.get(template_name)
+                if template is None:
+                    template = context.template.engine.get_template(template_name)
+                context.render_context[template_name] = template
```

Comment:
This is mainly to avoid edge cases. Here's an example:

Imagine a 3rd-party node sets a value with key ``include.html`` in ``context.render_context``. It's not probable, but perfectly possible. Later ``IncludeNode`` runs and happens to also use ``include.html`` as a key. It will pull whatever is already in ``render_context`` from that previous node when it shouldn't.

Using a custom key won't eliminate the edge case completely, but it does reduce the probability of edge cases. The ``ExtendsNode`` uses a key value of ``extends_context``. ``IncludeNode`` could do something similar.